### ISSUE TITLE
[BUG] Cannot enable a channel with a third party extension via Administrator console

### ISSUE THREAD

**Issue by marekhorban** [nextgenhealthcare/connect#6348]
**Describe the bug**
Mirth Administrator console doesn't work with a third party extension (connector), even when it is properly whitelisted as it states here: https://github.com/nextgenhealthcare/connect/wiki/4.4.1---Upgrade-Guide

**To Reproduce**
1. Configure allow types: in the mirth.properties: add line with "xstream.allowtypes = com.xyz.**"
2. Restart the Mirth Connect server
3. Run Administration console, go the "Extensions" and install a connector that has all classes in the package com.xyz.**
4. Create new channel using the connector as source, save the channel and try to deploy
5. Go to channels again: you see the created channel disabled with Description: "This channel is invalid. Verify all required extensions are loaded correctly." The channel cannot be edited or enabled.
6. When you click on the channel a stacktrace like this will be presented:
`Channel "xxx" is invalid and cannot be edited. Original cause:
com.xyz.RabbitMQReceiverProperties
com.thoughtworks.xstream.security.ForbiddenClassException: com.xyz.RabbitMQReceiverProperties
at com.thoughtworks.xstream.security.NoTypePermission.allows(NoTypePermission.java:26)
at com.thoughtworks.xstream.mapper.SecurityMapper.realClass(SecurityMapper.java:74)
at com.thoughtworks.xstream.mapper.MapperWrapper.realClass(MapperWrapper.java:125)
at com.thoughtworks.xstream.mapper.CachingMapper.realClass(CachingMapper.java:47)
at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:420)`

**Expected behavior**
A channel using a third party connector with whitelisted package works using Administrator console.

**Actual behavior**
A channel using a third party connector with whitelisted package is disabled and throws an error in the Administrator console

**Environment (please complete the following information):**
 - OS: Windows 11
 - Java Distribution/Version: OpenJDK 17,
 - Connect Version: 4.5.0

**Workaround(s)**
Not to use the admin console

**Additional context**
The bug exists starting from version 4.4.1 where this issue was resolved: https://github.com/nextgenhealthcare/connect/issues/5952 
Currently both server and client use xstream in version 1.4.20 which denies all classes which are not whitelisted by default. On server side the whitelisting configuration is handled but the client configuration is completely forgotten.
This issue introduces a strange behavior. When we click "Redeploy all" on the disabled channel it starts and works, sic! (server whitelisting allows to that) we can see it on the dashboard, but still on the channels tab the channel is disabled and cannot be enabled. This is kind of inconsistent state of the application.

**Comment by jonbartels** [comment]
Take a look at how the `xstream.allowtypes` property is used: https://github.com/nextgenhealthcare/connect/blob/be90435c57f2f0e93f1aa612f5afc4bf52717e01/server/src/com/mirth/connect/server/controllers/DefaultConfigurationController.java#L429

Does your plugin need to make similar calls in its own *client side* code to tell the Serializer to trust the objects that the plugin needs to serialize/deserialize?

**Comment by kayyagari** [comment]
The client part of plugin is likely not using `com.mirth.connect.model.converters.ObjectXMLSerializer`, which understands `xstream.allowtypes`.

**Comment by marekhorban** [comment]
> Take a look at how the `xstream.allowtypes` property is used:
> 
> https://github.com/nextgenhealthcare/connect/blob/be90435c57f2f0e93f1aa612f5afc4bf52717e01/server/src/com/mirth/connect/server/controllers/DefaultConfigurationController.java#L429
> 
> Does your plugin need to make similar calls in its own _client side_ code to tell the Serializer to trust the objects that the plugin needs to serialize/deserialize?

The `DefaultConfigurationController` is on server side and it works well. It doesn't exist on client side. Additionally plugin shouldn't be able to whitelist itself (from the plugin code). That possibility enables again the vulnerability: https://nvd.nist.gov/vuln/detail/CVE-2023-43208.
Whitelisting should be only possible by external configuration by admin, as it works on server side, but doesn't work on client side.

**Comment by kayyagari** [comment]
> Whitelisting should be only possible by external configuration by admin, as it works on server side, but doesn't work on client side.

Also by the developer of the plugin, by explicitly whitelisting the relevant classes/packages in the code.

**Comment by marekhorban** [comment]
> > Whitelisting should be only possible by external configuration by admin, as it works on server side, but doesn't work on client side.
> 
> Also by the developer of the plugin, by explicitly whitelisting the relevant classes/packages in the code.

Ok, thank you for the information. 
Our plugin fully worked on server and client sides before 4.4.1. Based on this doc: https://github.com/nextgenhealthcare/connect/wiki/4.4.1---Upgrade-Guide we've added described configuration to the mirth.properties. If there is an additional configuration which is not covered in mentioned link please point out the documentation or briefly describe what needs to be configured additionally.

**Comment by marekhorban** [comment]
@kayyagari @jonbartels 
The issue is still blocking us.
Since no action has been taken on this, I would propose preparing a pull request to fix the bug myself, unless you could address it instead?

Since the Mirth client does not handle any configuration files (unlike the server side: `mirth.properties`), I propose using the `launcher.vmoptions` to whitelist plugin packages. An example configuration entry could be: `-Dxstream.allowtypes=com.xyz.**`. This VM property would then be handled similarly to how it is on the server side. Would you agree with this approach?

And a final question: when can we expect the new version of Mirth to be released once the fix is approved and merged?

**Comment by jonbartels** [comment]
Neither Kiran nor I work for Nextgen. We are open-source Mirth users just like you!

If you can craft a PR to fix the issue, you should make one. It helps the community.

Did you try the client side whitelisting in your client side plugin code? Did it work?

> And a final question: when can we expect the new version of Mirth to be released once the fix is approved and merged?

Take a look at the existing PR backlog. NextGen releases Mirth updates quarterly but does not always incorporate every PR from the community in each release.

**Comment by marekhorban** [comment]
@jonbartels 
Thank you for your input. 
I had tried the client side whitelisting earlier but without any effect, but I didn't test every option. Now when you encouraged me to do it again, I finally find a way to do it successfully!

I thought "this cannot be possible" as whitelisting itself seems to be a security issue - precisely missing external whitelisting is a security issue. Now we have situation that malicious code on client side can grant itself to execute. There is no option to deny it on client side.
Nonetheless it is another story. 
I'm really happy that now we can use our plugin. In a free time I will create a documentation ticket to update migration documentation.
I'm closing this ticket.

**Comment by tonygermano** [comment]
> Thank you for your input. I had tried the client side whitelisting earlier but without any effect, but I didn't test every option. Now when you encouraged me to do it again, I finally find a way to do it successfully!

Can you share what you found that worked successfully to resolve your issue? It may help future developers and prevent this same ticket from being opened again by someone else.

> I thought "this cannot be possible" as whitelisting itself seems to be a security issue - precisely missing external whitelisting is a security issue. Now we have situation that malicious code on client side can grant itself to execute.

The reason for whitelisting is to prevent remote execution, but if you've already allowed a malicious plugin to run, then there is no reason for trying to enable an xstream exploit. From the client side, whitelisting a dangerous class would also require the mirth server itself to execute the attack on the client, which again, seems like a lot of wasted effort if the user already allowed the plugin code to run in the first place.
